### ISSUE TITLE
can not use image with absolute url

### ISSUE THREAD

**Issue by kaedea** [LouisBarranqueiro/tranquilpeak-hexo-theme#101]
I have got a problem when using image is the post. When I use image from asset folder it works well. While it works wrong when I use outter image with absoute url because the final image in the post is wrong.

![enter image description here](https://lh3.googleusercontent.com/-jxL1KWNuo_Y/VglHk-SB_8I/AAAAAAAABOY/nPcPYHzQfqw/s1000/BaiduShurufa_2015-9-28_21-57-25.jpg "BaiduShurufa_2015-9-28_21-57-25.jpg")

Is it that I mistake something in the setting?

**Comment by h404bi** [comment]
Change the [`url` value](https://github.com/kaedea/kaede-hexo/blob/master/_config.yml#L15) of your hexo `_config.yml` to your site domain name(i.e. `http://www.kaedea.com`).

**Comment by Imyssed** [comment]
Check the value 'url' and 'root' in _config.yml of Hexo

**Comment by kaedea** [comment]
When I set a concrete domain to the `url`, it works but the urls of the thumbnail images of the post will also be generated with this concrete domain. Therefore if I run `Hexo s` to deploy the website in localhost, these images' url will be generated as `domain + relative url`, which are supposed to be `localhost +  relative url`.
But this problem will be solved when the website is deployed to remote host. 
Anyway, THKS.

**Comment by LouisBarranqueiro** [comment]
@kaedea could you show me the source code of your article and hexo configuration file please. And which version of Tranquilpeak do you have?

**Comment by h404bi** [comment]
I saw that the article is from the demo site, and theme is v1.3.0.
And what I suggested above(edit the url) is just a workaround, because this will cause a issue mentioned above.
Therefore, you should use localhost as `url` when you debugging in localhost, and domain name as `url` when deploying to remote host. This is the way i used. Annoying. :(

**Comment by kaedea** [comment]
@LouisBarranqueiro 
The problem just as @h404bi said. But this only appears in the post thumbnail image. The image inserted in the post works well.

**Comment by LouisBarranqueiro** [comment]
You don't have to edit the `config.url` parameter. You can let the remote url. Its works well for me :)

**Comment by h404bi** [comment]
@LouisBarranqueiro are you sure? If I set `config.url` to `/`, like @kaedea , `front-matter` images(coverImage, thumbnailImage) url will generate wrongly.

**Comment by kaedea** [comment]
@LouisBarranqueiro
The Cover Image does `not` have this problem.

**Comment by h404bi** [comment]
@kaedea `coverImage` in `front-matter` has, not site cover image. I have tested it before.

**Comment by h404bi** [comment]
`photos` variable in `front-matter` also. Perhaps I found the reason, I'm trying to fix this issue.

**Comment by LouisBarranqueiro** [comment]
Of course, you have to provide a correct url! but you can let the remote url for dev environment.

**Comment by h404bi** [comment]
Set `config.url` to `/`, so I don't have to set it to `http://localhost:4000` when debugging in localhost, and set it to `http://example.com` when deploying. Every time after debugging in localhost, I have to change the `config.url` back to `http://example.com` to prevent it from being pushed to remote host. But never mind.

- - -

@kaedea If you still want to use `/` as `config.url`, you can edit these source files:
``` diff
modified:   layout/_partial/index.ejs
modified:   layout/_partial/post/gallery.ejs
modified:   layout/_partial/post/header-cover.ejs
```
``` diff
diff --git a/layout/_partial/index.ejs b/layout/_partial/index.ejs
index 6db97e9..24431d9 100644
--- a/layout/_partial/index.ejs
+++ b/layout/_partial/index.ejs
@@ -2,19 +2,19 @@
 <section class="postShorten-group main-content-wrap">
     <% page.posts.each(function(post) { %>
         <% if ((theme.thumbnail_image) && (post.thumbnailImage) && ((post.thumbnailIma                                                                                                                                                        ge.length))) { %>
-            <% if ((post.thumbnailImage.indexOf(config.url) < 0) && (post.thumbnailIma                                                                                                                                                        ge.indexOf('://') >= 0)) { %>
+            <% if ((post.thumbnailImage.indexOf(config.url) < 0) || (post.thumbnailIma                                                                                                                                                        ge.indexOf('://') >= 0)) { %>
                 <% thumbnailImageUrl = post.thumbnailImage; %>
             <% } else { %>
                 <% thumbnailImageUrl = url_for(post.permalink + post.thumbnailImage);                                                                                                                                                         %>
             <% } %>
         <% } else if ((theme.thumbnail_image) && (post.photos) && (post.photos.length)                                                                                                                                                        ) { %>
-            <% if ((post.photos[0].indexOf(config.url) < 0) && (post.photos[0].indexOf                                                                                                                                                        ('://') >= 0)) { %>
+            <% if ((post.photos[0].indexOf(config.url) < 0) || (post.photos[0].indexOf                                                                                                                                                        ('://') >= 0)) { %>
:...skipping...
diff --git a/layout/_partial/index.ejs b/layout/_partial/index.ejs
index 6db97e9..24431d9 100644
--- a/layout/_partial/index.ejs
+++ b/layout/_partial/index.ejs
@@ -2,19 +2,19 @@
 <section class="postShorten-group main-content-wrap">
     <% page.posts.each(function(post) { %>
         <% if ((theme.thumbnail_image) && (post.thumbnailImage) && ((post.thumbnailImage.length))) { %>
-            <% if ((post.thumbnailImage.indexOf(config.url) < 0) && (post.thumbnailImage.indexOf('://') >= 0)) { %>
+            <% if ((post.thumbnailImage.indexOf(config.url) < 0) || (post.thumbnailImage.indexOf('://') >= 0)) { %>
                 <% thumbnailImageUrl = post.thumbnailImage; %>
             <% } else { %>
                 <% thumbnailImageUrl = url_for(post.permalink + post.thumbnailImage); %>
             <% } %>
         <% } else if ((theme.thumbnail_image) && (post.photos) && (post.photos.length)) { %>
-            <% if ((post.photos[0].indexOf(config.url) < 0) && (post.photos[0].indexOf('://') >= 0)) { %>
+            <% if ((post.photos[0].indexOf(config.url) < 0) || (post.photos[0].indexOf('://') >= 0)) { %>
                 <% thumbnailImageUrl = post.photos[0]; %>
             <% } else { %>
                 <% thumbnailImageUrl = url_for(post.permalink + post.photos[0]); %>
             <% } %>
         <% } else if ((theme.thumbnail_image) && (post.coverImage) && (post.coverImage.length)) { %>
-            <% if ((post.coverImage.indexOf(config.url) < 0) && (post.coverImage.indexOf('://') >= 0)) { %>
+            <% if ((post.coverImage.indexOf(config.url) < 0) || (post.coverImage.indexOf('://') >= 0)) { %>
                 <% thumbnailImageUrl = post.coverImage; %>
             <% } else { %>
                 <% thumbnailImageUrl = url_for(post.permalink + post.coverImage); %>
diff --git a/layout/_partial/post/gallery.ejs b/layout/_partial/post/gallery.ejs
index f62783c..e8654fc 100644
--- a/layout/_partial/post/gallery.ejs
+++ b/layout/_partial/post/gallery.ejs
@@ -13,7 +13,7 @@
                     <% var match = photo.match(rPhoto); %>
                     <% var photoUrl = match[1]; %>
                     <% var photoTitle = match[3] || ""; %>
-                <% if ((photoUrl.indexOf(config.url) < 0) && (photoUrl.indexOf('://') >= 0)) { %>
+                <% if ((photoUrl.indexOf(config.url) < 0) || (photoUrl.indexOf('://') >= 0)) { %>
                     <%  photoUrl = photoUrl; %>
                     <% if (!/.+\.[a-zA-Z0-9]{3,4}$/.test(photoUrl)) { %>
                         <%  photoUrl += "#.jpg"; %>
diff --git a/layout/_partial/post/header-cover.ejs b/layout/_partial/post/header-cover.ejs
index f4f0411..b480d9b 100644
--- a/layout/_partial/post/header-cover.ejs
+++ b/layout/_partial/post/header-cover.ejs
@@ -3,7 +3,7 @@
     var coverImageUrl;

     if (post.coverImage) {
-        if ((post.coverImage.indexOf(config.url) < 0) && (post.coverImage.indexOf('://') >= 0)) {
+        if ((post.coverImage.indexOf(config.url) < 0) || (post.coverImage.indexOf('://') >= 0)) {
             coverImageUrl = post.coverImage;
         }
         else {
~
~
```
Then it would work correctly, I have tested it. (**workaround**)

**Comment by kaedea** [comment]
@h404bi
thks, i am trying this.

**Comment by LouisBarranqueiro** [comment]
I did some test and what you have forget to say @kaedea is that you have set `relative_link` parameter to `true`. and the when you set `config.url` to `/`, yes it generate a bug only because you incorrectly set the `config.url` parameter. There is actually no bugs. You are currently lazy like @h404bi :p and you just need a feature to not to have to change the `config.url` parameter each time you deploy and write a post. Am i right?

**Comment by LouisBarranqueiro** [comment]
about the code @h404bi, it can be improved, i will do that in a PR

**Comment by h404bi** [comment]
Well, `relative_link` is `false`. And, haha, in fact I have been used to it, never mind. And I have no any advice about the PR, up to you. ;-)

**Comment by Imyssed** [comment]
When using `asset_folder`,don't use`/`,use `image1.png`,etc

**Comment by kaedea** [comment]
I think I have to point out that the problem I want to say is:
Now the thumbnail image's url will be generate as a absolute url in final html file (e.x. `<img alt="" itemprop="image" src="http://localhost:4000/2015/09/29/themia-demo-cover-image/post-thumb.jpg" style="display: block;">`). The `url` parameter will be added. However, this means I need to set the `url` to `localhost` when debuging locally and then set it to `mydomain` when deploying to remote.
Why not work like the coverimage. It will be generated as a relative url in the html file (e.x ` style="background-image:url('/2015/09/29/themia-demo-cover-image/post-cover.jpg')";`). I do not need to change the  `url` parameter any more.

**Comment by Imyssed** [comment]
:joy:Not ever met this bug.

**Comment by h404bi** [comment]
@kaedea 
Conclusion: you want `thumbnailImageUrl` to be generated as a relative url.
- - -
In [layout/_partial/index.ejs#L8](https://github.com/LouisBarranqueiro/tranquilpeak-hexo-theme/blob/master/layout/_partial/index.ejs#L8), `thumbnailImageUrl` use `post.permalink`, which is a full URL of the article([^1]), while in [layout/_partial/post/header-cover.ejs#L10](https://github.com/LouisBarranqueiro/tranquilpeak-hexo-theme/blob/master/layout/_partial/post/header-cover.ejs#L10), `coverImageUrl` use `post.path`.

So, this may suit you:
``` diff
diff --git a/layout/_partial/index.ejs b/layout/_partial/index.ejs
index 6db97e9..3a75d0f 100644
--- a/layout/_partial/index.ejs
+++ b/layout/_partial/index.ejs
@@ -2,22 +2,22 @@
 <section class="postShorten-group main-content-wrap">
     <% page.posts.each(function(post) { %>
         <% if ((theme.thumbnail_image) && (post.thumbnailImage) && ((post.thumbnailImage.length))) { %>
-            <% if ((post.thumbnailImage.indexOf(config.url) < 0) && (post.thumbnailImage.indexOf('://') >= 0)) { %>
+            <% if (post.thumbnailImage.indexOf('://') >= 0) { %>
                 <% thumbnailImageUrl = post.thumbnailImage; %>
             <% } else { %>
-                <% thumbnailImageUrl = url_for(post.permalink + post.thumbnailImage); %>
+                <% thumbnailImageUrl = url_for(post.path.replace('.html', '/') + post.thumbnailImage); %>
             <% } %>
         <% } else if ((theme.thumbnail_image) && (post.photos) && (post.photos.length)) { %>
-            <% if ((post.photos[0].indexOf(config.url) < 0) && (post.photos[0].indexOf('://') >= 0)) { %>
+            <% if (post.photos[0].indexOf('://') >= 0) { %>
                 <% thumbnailImageUrl = post.photos[0]; %>
             <% } else { %>
-                <% thumbnailImageUrl = url_for(post.permalink + post.photos[0]); %>
+                <% thumbnailImageUrl = url_for(post.path.replace('.html', '/') + post.photos[0]); %>
             <% } %>
         <% } else if ((theme.thumbnail_image) && (post.coverImage) && (post.coverImage.length)) { %>
-            <% if ((post.coverImage.indexOf(config.url) < 0) && (post.coverImage.indexOf('://') >= 0)) { %>
+            <% if (post.coverImage.indexOf('://') >= 0) { %>
                 <% thumbnailImageUrl = post.coverImage; %>
             <% } else { %>
-                <% thumbnailImageUrl = url_for(post.permalink + post.coverImage); %>
+                <% thumbnailImageUrl = url_for(post.path.replace('.html', '/') + post.coverImage); %>
             <% } %>
         <% } %>
         <% if (thumbnailImageUrl != null) { %>
diff --git a/layout/_partial/post/gallery.ejs b/layout/_partial/post/gallery.ejs
index f62783c..51b9a6c 100644
--- a/layout/_partial/post/gallery.ejs
+++ b/layout/_partial/post/gallery.ejs
@@ -13,7 +13,7 @@
                     <% var match = photo.match(rPhoto); %>
                     <% var photoUrl = match[1]; %>
                     <% var photoTitle = match[3] || ""; %>
-                <% if ((photoUrl.indexOf(config.url) < 0) && (photoUrl.indexOf('://') >= 0)) { %>
+                <% if (photoUrl.indexOf('://') >= 0) { %>
                     <%  photoUrl = photoUrl; %>
                     <% if (!/.+\.[a-zA-Z0-9]{3,4}$/.test(photoUrl)) { %>
                         <%  photoUrl += "#.jpg"; %>
diff --git a/layout/_partial/post/header-cover.ejs b/layout/_partial/post/header-cover.ejs
index f4f0411..d1be3e0 100644
--- a/layout/_partial/post/header-cover.ejs
+++ b/layout/_partial/post/header-cover.ejs
@@ -3,7 +3,7 @@
     var coverImageUrl;
 
     if (post.coverImage) {
-        if ((post.coverImage.indexOf(config.url) < 0) && (post.coverImage.indexOf('://') >= 0)) {
+        if (post.coverImage.indexOf('://') >= 0) {
             coverImageUrl = post.coverImage;
         }
         else {
```

:new_moon_with_face:

[^1]: https://hexo.io/docs/variables.html#Page_Variables

**Comment by LouisBarranqueiro** [comment]
fix that @h404bi on #105, your job! :D

**Comment by clkao** [comment]
It seems path resolution for thumbnailImage does not have the same logic as coverImage, making it impossible to refer to things in assets path unless using the full url.

Also when thumbnailImage is not defined, it uses coverImage when available. so if the coverImage is in assets path, you get a broken thumbnail image.  It'd be also nice to be able to disable thumbnail in this case by specifying "thumbnaiImage: false" for particular posts.

**Comment by LouisBarranqueiro** [comment]
It's already fixed for 1.4.0 and user will be able to disable or enable automatic selection of thirds image as thumbnail image. you can check it on the `dev` branch :)

**Comment by clkao** [comment]
Oh I did try the dev branch, how does one disable thumbnail for posts with cover image?

**Comment by LouisBarranqueiro** [comment]
set `autoThumbnailImage` to  `false`

**Comment by LouisBarranqueiro** [comment]
I just released the new version [1.4.0](https://github.com/LouisBarranqueiro/tranquilpeak-hexo-theme/releases/tag/v1.4.0), read user doc - [writing posts -> front-matter settings](https://github.com/LouisBarranqueiro/tranquilpeak-hexo-theme/blob/master/docs/user.md#front-matter-settings) and [theme configuration -> customization](https://github.com/LouisBarranqueiro/tranquilpeak-hexo-theme/blob/master/docs/user.md#customization) sections

You should be able to do what you want now. Re-open the issue, if you encounter some problem
